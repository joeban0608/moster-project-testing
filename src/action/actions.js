import { apiCall } from "../api/api";
import {
  CHANGE_SEARCHFIELD,
  REQUEST_ROBOTS_PENDING,
  REQUEST_ROBOTS_SUCCESS,
  REQUEST_ROBOTS_FAILED,
} from "./constants";

export const setSearchField = (text) => ({
  type: CHANGE_SEARCHFIELD,
  payload: text,
});

export const fetchRobotsSuccess = (robots) => ({
  type: REQUEST_ROBOTS_SUCCESS,
  payload: robots,
});
export const fetchRobotsFailed = (error) => ({
  type: REQUEST_ROBOTS_FAILED,
  payload: error,
});
export const fetchRobotsStart = () => ({
  type: REQUEST_ROBOTS_PENDING,
});

export const setRequestRobots = () => (dispatch) => {
  dispatch(fetchRobotsStart());
  apiCall("https://jsonplaceholder.typicode.com/users")
    .then((robots) => dispatch(fetchRobotsSuccess(robots)))
    .catch((error) => dispatch(fetchRobotsFailed(error)));
};
