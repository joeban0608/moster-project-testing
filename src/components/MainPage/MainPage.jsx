import React, { useEffect, useState } from "react";
import CardList from "../card-list/card-list.component";
import CounterButton from "../counter-button/counter-button.component";
import SearchBox from "../search-box/search-box.component";
import "./MainPage.css";

const MainPage = ({ ...props }) => {
  const { robots, searchField, onSearchChange, onRequestMonsters } = props;
  const [filteredMonsters, setFilterMonsters] = useState(robots);
  useEffect(() => {
    onRequestMonsters();
  }, []);

  const handleSetFilterMonsters = () => {
    const newFilteredMonsters = robots.filter((monster) => {
      return monster.name.toLocaleLowerCase().includes(searchField);
    });

    setFilterMonsters(newFilteredMonsters);
  };

  useEffect(() => {
    handleSetFilterMonsters();
  }, [robots, searchField]);

  return (
    <div className="main">
      <h1 className="main-title">Monsters Rolodex</h1>
      <CounterButton />
      <SearchBox
        className="monsters-search-box"
        onChangeHandler={onSearchChange}
        placeholder="search monsters"
      />
      <CardList monsters={filteredMonsters} />
    </div>
  );
};

export default MainPage;
