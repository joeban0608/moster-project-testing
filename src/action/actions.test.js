import * as actions from "./actions";
import {
  CHANGE_SEARCHFIELD,
  REQUEST_ROBOTS_PENDING,
  REQUEST_ROBOTS_SUCCESS,
  REQUEST_ROBOTS_FAILED,
} from "./constants";

const {
  setSearchField,
  setRequestRobots,
  fetchRobotsSuccess,
  fetchRobotsFailed,
  fetchRobotsStart,
} = actions;

import configureMockStore from "redux-mock-store";
import thunkMiddleware from "redux-thunk";
import { mockRobots } from "../components/MainPage/MainPage.test";

const mockStore = configureMockStore([thunkMiddleware]);

describe("handle setSearchField action", () => {
  it("should create an action to search robots", () => {
    const text = "ervin";
    const expectedAction = {
      type: CHANGE_SEARCHFIELD,
      payload: text,
    };
    expect(setSearchField(text)).toEqual(expectedAction);
  });
});

describe("handle robots request Api", () => {
  it("handle REQUEST_ROBOTS_PENDING action 1", () => {
    const store = mockStore();
    store.dispatch(setRequestRobots());
    const action = store.getActions();
    // console.log('action', action)
    const expectedAction = {
      type: REQUEST_ROBOTS_PENDING,
    };
    expect(action[0]).toEqual(expectedAction);
  });

  it("handle fetchRobotsStart", () => {
    expect(fetchRobotsStart().type).toEqual(REQUEST_ROBOTS_PENDING);
  });

  it("should create fetchRobotsSuccess action", () => {
    const action = fetchRobotsSuccess(mockRobots);
    expect(action.type).toEqual(REQUEST_ROBOTS_SUCCESS);
    expect(action.payload).toEqual(mockRobots);
  });

  it("handle fetchRobotsFailed action", () => {
    const action = fetchRobotsFailed(Error);
    expect(action.type).toEqual(REQUEST_ROBOTS_FAILED);
    expect(action.payload).toEqual(Error);
  });

  it("another way handle setRequestRobots action", () => {
    const mockActionCreator = setRequestRobots();
    const mockDispatch = jest.fn();
    mockActionCreator(mockDispatch);
    expect(mockDispatch).toHaveBeenCalledWith(fetchRobotsStart());
  });
});
