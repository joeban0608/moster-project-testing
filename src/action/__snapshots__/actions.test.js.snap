// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`is render MainPage component 1`] = `
<div
  className="main"
>
  <h1
    className="main-title"
  >
    Monsters Rolodex
  </h1>
  <CounterButton />
  <SearchBox
    className="monsters-search-box"
    placeholder="search monsters"
  />
  <CardList
    monsters={Array []}
  />
</div>
`;
